src/types/coefficient.rs

```diff
@@ -0,0 +1,133 @@
+use num_bigint::{BigUint, BigInt};
+use num_traits::{Zero};
+
+use crate::types::magnitude::Magnitude;
+use std::ops::MulAssign;
+
+/// Indicates whether the Coefficient's magnitude is less than 0 (negative) or not (positive).
+/// When the magnitude is zero, the Sign can be used to distinguish between -0 and 0.
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Debug)]
+pub enum Sign {
+    Negative,
+    Positive,
+}
+
+/// A signed integer that can be used as the coefficient of a Decimal value. This type does not
+/// consider `0` and `-0` to be equal and supports magnitudes of arbitrary size.
+// These trait derivations rely closely on the manual implementations of PartialEq and Ord on
+// [Magnitude].
+#[derive(Debug, Clone, Ord, PartialOrd, Eq, PartialEq)]
+pub struct Coefficient {
+    sign: Sign,
+    magnitude: Magnitude,
+}
+
+impl Coefficient {
+    pub(crate) fn new<I: Into<Magnitude>>(sign: Sign, magnitude: I) -> Self {
+        let magnitude = magnitude.into();
+        Coefficient {
+            sign,
+            magnitude,
+        }
+    }
+
+    pub(crate) fn sign(&self) -> Sign {
+        self.sign
+    }
+
+    pub(crate) fn magnitude(&self) -> &Magnitude {
+        &self.magnitude
+    }
+
+    pub(crate) fn negative_zero() -> Self {
+        Coefficient {
+            sign: Sign::Negative,
+            magnitude: Magnitude::U64(0)
+        }
+    }
+
+    pub(crate) fn is_negative_zero(&self) -> bool {
+        match (self.sign, &self.magnitude) {
+            (Sign::Negative, Magnitude::U64(0)) => true,
+            (Sign::Negative, Magnitude::BigUInt(b)) if b.is_zero() => true,
+            _ => false
+        }
+    }
+}
+
+// This macro makes it possible to turn unsigned integers into a Coefficient using `.into()`.
+macro_rules! impl_coefficient_from_unsigned_int_types {
+    ($($t:ty),*) => ($(
+        impl From<$t> for Coefficient {
+            fn from(value: $t) -> Coefficient {
+                Coefficient::new(Sign::Positive, value)
+            }
+        }
+    )*)
+}
+impl_coefficient_from_unsigned_int_types!(u8, u16, u32, u64, u128, usize, BigUint);
+
+// This macro makes it possible to turn signed integers into a Coefficient using `.into()`.
+macro_rules! impl_coefficient_from_signed_int_types {
+    ($($t:ty),*) => ($(
+        impl From<$t> for Coefficient {
+            fn from(value: $t) -> Coefficient {
+                let sign = if value < <$t>::zero() { Sign::Negative } else { Sign::Positive };
+                Coefficient::new(sign, value)
+            }
+        }
+    )*)
+}
+impl_coefficient_from_signed_int_types!(i8, i16, i32, i64, i128, isize);
+
+// `BigInt` can't represent -0, so this is technically a lossy operation.
```

Comment:
Ah that old chestnut again. Is the plan to "wrap" the type and track the `-0` case?

Reply:
I've replaced the `From<Decimal> for BigDecimal` and `From<Coefficient> for BigInt` implementations with `TryFrom` implementations that fail if the value being converted is a negative zero (and so would cause data loss).

If we need one, we can add a `From<Decimal> for (Sign, BigDecimal)` or other option for guaranteed conversion.